src/dump.c

```diff
@@ -447,68 +459,99 @@ static int is_ast_node(jl_value_t *v)
 
 static int literal_val_id(jl_serializer_state *s, jl_value_t *v)
 {
-    int i, l = jl_array_len(s->tree_literal_values);
-    for (i = 0; i < l; i++) {
-        if (jl_egal(jl_array_ptr_ref(s->tree_literal_values, i), v))
-            return i;
+    jl_array_t *rs = s->method->roots;
+    int i, l = jl_array_len(rs);
+    if (jl_is_symbol(v) || jl_is_concrete_type(v)) {
+        for (i = 0; i < l; i++) {
+            if (jl_array_ptr_ref(rs, i) == v)
+                return i;
+        }
+    }
+    else {
+        for (i = 0; i < l; i++) {
+            if (jl_egal(jl_array_ptr_ref(rs, i), v))
+                return i;
+        }
     }
-    jl_array_ptr_1d_push(s->tree_literal_values, v);
-    return jl_array_len(s->tree_literal_values) - 1;
+    jl_array_ptr_1d_push(rs, v);
+    return jl_array_len(rs) - 1;
 }
 
 static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_literal)
 {
     if (v == NULL) {
-        write_uint8(s->s, Null_tag);
+        write_uint8(s->s, TAG_NULL);
         return;
     }
 
-    void **bp = ptrhash_bp(&ser_tag, v);
-    if (*bp != HT_NOTFOUND) {
-        write_as_tag(s->s, (uint8_t)(intptr_t)*bp);
+    void *tag = ptrhash_get(&ser_tag, v);
```

Comment:
If this is performance-sensitive at all, I think we could guard it these `ptrhash_get` with `((jl_datatype_t*)jl_typeof(ser_tag))->name->module == jl_core_module` I'm not sure that would make any difference however, but just an idea.